orchagent/fabricportsorch.cpp

```diff
@@ -0,0 +1,268 @@
+#include "fabricportsorch.h"
+
+#include <inttypes.h>
+#include <fstream>
+#include <sstream>
+#include <tuple>
+
+#include "logger.h"
+#include "schema.h"
+#include "sai_serialize.h"
+#include "timer.h"
+
+#define FABRIC_POLLING_INTERVAL_DEFAULT   (30)
+#define FABRIC_PORT_ERROR     0
+#define FABRIC_PORT_SUCCESS   1
+#define FABRIC_PORT_STAT_COUNTER_FLEX_COUNTER_GROUP         "FABRIC_PORT_STAT_COUNTER"
+#define FABRIC_PORT_STAT_FLEX_COUNTER_POLLING_INTERVAL_MS   10000
+#define FABRIC_QUEUE_STAT_COUNTER_FLEX_COUNTER_GROUP        "FABRIC_QUEUE_STAT_COUNTER"
+#define FABRIC_QUEUE_STAT_FLEX_COUNTER_POLLING_INTERVAL_MS  100000
+#define FABRIC_PORT_TABLE "FABRIC_PORT_TABLE"
+
+extern sai_object_id_t gSwitchId;
+extern sai_switch_api_t *sai_switch_api;
+extern sai_port_api_t *sai_port_api;
+
+const vector<sai_port_stat_t> port_stat_ids =
+{
+    SAI_PORT_STAT_IF_IN_OCTETS,
+    SAI_PORT_STAT_IF_IN_ERRORS,
+    SAI_PORT_STAT_IF_IN_FABRIC_DATA_UNITS,
+    SAI_PORT_STAT_IF_IN_FEC_CORRECTABLE_FRAMES,
+    SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES,
+    SAI_PORT_STAT_IF_IN_FEC_SYMBOL_ERRORS,
+    SAI_PORT_STAT_IF_OUT_OCTETS,
+    SAI_PORT_STAT_IF_OUT_FABRIC_DATA_UNITS,
+};
+
+static const vector<sai_queue_stat_t> queue_stat_ids =
+{
+    SAI_QUEUE_STAT_WATERMARK_LEVEL,
+    SAI_QUEUE_STAT_CURR_OCCUPANCY_BYTES,
+    SAI_QUEUE_STAT_CURR_OCCUPANCY_LEVEL,
+};
+
+FabricPortsOrch::FabricPortsOrch(DBConnector *appl_db, vector<table_name_with_pri_t> &tableNames) :
+        Orch(appl_db, tableNames),
+        port_stat_manager(FABRIC_PORT_STAT_COUNTER_FLEX_COUNTER_GROUP, StatsMode::READ,
+                          FABRIC_PORT_STAT_FLEX_COUNTER_POLLING_INTERVAL_MS, true),
+        queue_stat_manager(FABRIC_QUEUE_STAT_COUNTER_FLEX_COUNTER_GROUP, StatsMode::READ,
+                           FABRIC_QUEUE_STAT_FLEX_COUNTER_POLLING_INTERVAL_MS, true),
+        m_timer(new SelectableTimer(timespec { .tv_sec = FABRIC_POLLING_INTERVAL_DEFAULT, .tv_nsec = 0 }))
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_NOTICE( "FabricPortsOrch constructor" );
+
+    m_state_db = shared_ptr<DBConnector>(new DBConnector("STATE_DB", 0));
+    m_stateTable = unique_ptr<Table>(new Table(m_state_db.get(), FABRIC_PORT_TABLE));
+
+    m_counter_db = shared_ptr<DBConnector>(new DBConnector("COUNTERS_DB", 0));
+    m_laneQueueCounterTable = unique_ptr<Table>(new Table(m_counter_db.get(), COUNTERS_QUEUE_NAME_MAP));
```

Comment:
what does COUNTERS_QUEUE_NAME_MAP and COUNTERS_QUEUE_PORT_MAP looks like for fabric port ?
Can you add comment in the description.

Reply:
Once jupiter-sai supports port and queue stats completely for fabric ports on both fabric and voq asic, this will look like below in `COUNTERS_DB`:
```
127.0.0.1:6379[2]> select 2
OK
```
```
127.0.0.1:6379[2]> HGETALL "COUNTERS_QUEUE_PORT_MAP"
  1) "FabricPort0"
  2) "oid:0x401000000000006"
```
```
127.0.0.1:6379[2]> HGETALL "COUNTERS:oid:0x401000000000006"
 1) "SAI_PORT_STAT_IF_OUT_FABRIC_DATA_UNITS"
 2) "0"
 3) "SAI_PORT_STAT_IF_IN_ERRORS"
 4) "0"
 5) "SAI_PORT_STAT_IF_IN_OCTETS"
 6) "0"
 7) "SAI_PORT_STAT_IF_IN_FABRIC_DATA_UNITS"
 8) "0"
 9) "SAI_PORT_STAT_IF_IN_FEC_SYMBOL_ERRORS"
10) "0"
11) "SAI_PORT_STAT_IF_OUT_OCTETS"
12) "0"
13) "SAI_PORT_STAT_IF_IN_FEC_CORRECTABLE_FRAMES"
14) "0"
15) "SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES"
16) "0"
```
```
127.0.0.1:6379[2]> HGETALL "COUNTERS_QUEUE_NAME_MAP"
  1) "FabricPort0:0"
  2) "oid:0x4150000000000c6"
```
```
127.0.0.1:6379[2]> HGETALL "COUNTERS:oid:0x4150000000000c6"
1) "SAI_QUEUE_STAT_CURR_OCCUPANCY_LEVEL"
2) "0"
3) "SAI_QUEUE_STAT_WATERMARK_LEVEL"
4) "24"
5) "SAI_QUEUE_STAT_CURR_OCCUPANCY_BYTES"
6) "0"
```